src/goto-symex/symex_goto.cpp

```diff
@@ -65,6 +67,103 @@ void goto_symext::apply_goto_condition(
   }
 }
 
+/// Try to evaluate pointer comparisons where they can be trivially determined
+/// using the value-set.
+/// \param condition: An expression with boolean type
+/// \param value_set: The value-set for determining what pointer-typed symbols
+///   might possibly point to
+/// \return The possibly modified condition
+exprt goto_symext::try_evaluate_pointer_comparisons(
+  exprt condition,
+  const value_sett &value_set)
+{
+  if(condition.id() == ID_and || condition.id() == ID_or)
+  {
+    for(auto &op : condition.operands())
+    {
+      op = try_evaluate_pointer_comparisons(std::move(op), value_set);
+    }
+  }
+  else if(condition.id() == ID_not)
+  {
+    for(auto &op : condition.operands())
+    {
+      op = try_evaluate_pointer_comparisons(std::move(op), value_set);
+    }
+  }
+  else if(condition.id() == ID_equal || condition.id() == ID_notequal)
+  {
+    if(can_cast_type<pointer_typet>(condition.op0().type()))
+    {
+      symbol_exprt *symbol_expr =
+        expr_try_dynamic_cast<symbol_exprt>(condition.op0());
+      exprt &constant = condition.op1();
+
+      if(symbol_expr && goto_symex_is_constantt()(constant))
+      {
+        value_setst::valuest value_set_elements;
+        value_set.get_value_set(
+          to_ssa_expr(*symbol_expr).get_l1_object(), value_set_elements, ns);
+
+        INVARIANT(
+          constant.id() == ID_address_of || constant.id() == ID_NULL,
```

Comment:
Surprising! Null is usually represented as an `ID_constant` with value = NULL -- suspect you need to add tests for this